src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokenPsiElement.java

```diff
@@ -0,0 +1,152 @@
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.semanticTokens.viewProvider;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiNameIdentifierOwner;
+import com.intellij.psi.impl.source.tree.LeafPsiElement;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.features.LSPPsiElement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+/**
+ * A PSI element for a semantic token in a {@link LSPSemanticTokensFileViewProvider} file.
+ */
+public class LSPSemanticTokenPsiElement extends LSPPsiElement implements PsiNameIdentifierOwner {
+    private final LSPSemanticToken semanticToken;
+    private final boolean isFile;
+    private volatile LeafPsiElement node = null;
+    private volatile String text = null;
+
+    /**
+     * Creates a new PSI element for the provided semantic token.
+     *
+     * @param semanticToken the semantic token
+     */
+    LSPSemanticTokenPsiElement(@NotNull LSPSemanticToken semanticToken) {
+        super(semanticToken.getFile(), semanticToken.getTextRange());
+        this.isFile = Objects.equals(semanticToken.getFile().getTextRange(), semanticToken.getTextRange());
+        this.semanticToken = semanticToken;
+    }
+
+    @Override
+    public boolean isPhysical() {
+        // These do represent real text ranges in physical files
+        return true;
+    }
+
+    @Override
+    public boolean isValid() {
+        // Tie the validity of this element to that of its containing file
+        return semanticToken.getFile().isValid();
+    }
+
+    @Override
+    public boolean canNavigate() {
+        // References can be navigated
+        return semanticToken.getElementType() == LSPSemanticTokenElementType.REFERENCE;
+    }
+
+    @Override
+    @Nullable
+    public PsiElement getNameIdentifier() {
+        // If this is a declaration, return this element as the name identifier
+        return semanticToken.getElementType() == LSPSemanticTokenElementType.DECLARATION ? this : null;
+    }
+
+    @Override
+    @Nullable
+    public ASTNode getNode() {
+        // This is lazy-initialized because it needs the element text which is itself lazy-initialized
+        if (node == null) {
+            synchronized (this) {
+                if (node == null) {
+                    node = new LeafPsiElement(semanticToken.getElementType(), getText());
+                }
+            }
+        }
+
+        return node;
+    }
+
+    @Override
+    @NotNull
+    public String getText() {
+        // Optimization for full-file elements to avoid copying the full file text
+        if (isFile) return getContainingFile().getText();
```

Comment:
And here's that optimization for full-file elements to delegate `getText()` instead of copying it locally.